src/tech-record/TechRecordService.ts

```diff
@@ -0,0 +1,128 @@
+import { Service } from 'typedi';
+import { IWeightDetails } from '../models';
+import { VEHICLE_TYPES } from '../models/Enums';
+import { HTTPError } from '../models/HTTPError';
+import { ISearchResult, TechRecordGet, TechRecordType } from '../models/Types';
+import { TechRecordRepository } from './TechRecordRepository';
+
+@Service()
+export class TechRecordService {
+	constructor(private techRecordRepository: TechRecordRepository) {}
+
+	/**
+	 * Method for getting make and model based on the vehicle from a test-result
+	 * @param testResult - the testResult for which the tech record search is done for
+	 */
+	public getVehicleMakeAndModel = async (testResult: any) => {
+		const searchRes = await this.techRecordRepository.callSearchTechRecords(testResult.systemNumber);
+		const techRecord = await this.processGetCurrentProvisionalRecords(searchRes);
+		// Return bodyMake and bodyModel values for PSVs
+		return techRecord?.techRecord_vehicleType === VEHICLE_TYPES.PSV
+			? {
+					Make: (techRecord as TechRecordType<'psv'>).techRecord_chassisMake,
+					Model: (techRecord as TechRecordType<'psv'>).techRecord_chassisModel,
+				}
+			: {
+					Make: (techRecord as TechRecordType<'hgv' | 'trl'>).techRecord_make,
+					Model: (techRecord as TechRecordType<'hgv' | 'trl'>).techRecord_model,
+				};
+	};
+
+	private processGetCurrentProvisionalRecords = async <T extends TechRecordGet['techRecord_vehicleType']>(
+		searchResult: ISearchResult[]
+	): Promise<TechRecordType<T> | undefined> => {
+		if (searchResult) {
+			const processRecordsRes = this.groupRecordsByStatusCode(searchResult);
+			return processRecordsRes.currentCount !== 0
+				? this.techRecordRepository.callGetTechRecords(
+						processRecordsRes.currentRecords[0].systemNumber,
+						processRecordsRes.currentRecords[0].createdTimestamp
+					)
+				: processRecordsRes.provisionalCount === 1
+					? this.techRecordRepository.callGetTechRecords(
+							processRecordsRes.provisionalRecords[0].systemNumber,
+							processRecordsRes.provisionalRecords[0].createdTimestamp
+						)
+					: this.techRecordRepository.callGetTechRecords(
+							processRecordsRes.provisionalRecords[1].systemNumber,
+							processRecordsRes.provisionalRecords[1].createdTimestamp
+						);
+		} else {
+			await Promise.reject('Tech record Search returned nothing.');
+		}
+	};
```

Comment:
Just checking, what happens here if `searchResult` is an empty array? Do we need to guard against that? If so, maybe the method would look like

```
 private async processGetCurrentProvisionalRecords<T extends TechRecordGet['techRecord_vehicleType']>(
    searchResult: ISearchResult[]
  ): Promise<TechRecordType<T> | undefined> {
    if (!searchResult || searchResult.length === 0) {
      throw new Error('Tech record Search returned nothing.');
    }

    const { currentRecords, provisionalRecords } = this.groupRecordsByStatusCode(searchResult);

    const recordToProcess = 
      currentRecords[0] || 
      provisionalRecords[0] || 
      provisionalRecords[1];

    if (!recordToProcess) {
      throw new Error('No suitable record found for processing.');
    }

    return this.techRecordRepository.callGetTechRecords(
      recordToProcess.systemNumber,
      recordToProcess.createdTimestamp
    );
  }
}
```

Either way, I'm happy. What are your thoughts?

Reply:
If `searchResult` is empty, is should throw the 'Tech record Search returned nothing.' error. Though I'm not a fan of this, it's basically a lift and shift of the existing logic.

I haven't really tried to improve the code on a functional level because of the subsequent changes to testing that would make it harder to know if regressions have been introduced.

In my opinion, errors should be reserved for legitimate unexpected behaviours, whether this meets that criteria (amongst many other places) needs evaluating. Similarly the practice of returning empty collections when an outbound request is successful but returns a 404 should be introduced to make code paths easier to follow.

This is all for a later PR though ;)